src/kernel/boot.c

```diff
@@ -874,20 +874,39 @@ BOOT_CODE bool_t init_freemem(word_t n_available, const p_region_t *available,
     word_t size = calculate_rootserver_size(it_v_reg, extra_bi_size_bits);
     word_t max = rootserver_max_size_bits(extra_bi_size_bits);
     for (; i >= 0; i--) {
-        word_t next = i + 1;
+        /* Invariant: both i and (i + 1) are valid indices in ndks_boot.freemem. */
+        assert(i < ARRAY_SIZE(ndks_boot.freemem) - 1);
+        /* Invariant; the region at index i is the current candidate.
+         * Invariant: regions 0 up to (i - 1), if any, are additional candidates.
+         * Invariant: region (i + 1) is undefined, and should not be read.
+         * Invariant: regions above (i + 1), if any, are empty or too small to use.
+         * Invariant: defined regions (all except i + 1) are ordered,
+         *            non-overlapping and unallocated. */
+
+        /* Can we fit a suitably aligned chunk in the top end of region i? */
         pptr_t start = ROUND_DOWN(ndks_boot.freemem[i].end - size, max);
         if (start >= ndks_boot.freemem[i].start) {
+            /* Yes, region i is big enough, so use it. */
             create_rootserver_objects(start, it_v_reg, extra_bi_size_bits);
-            if (i < ARRAY_SIZE(ndks_boot.freemem)) {
-                ndks_boot.freemem[next].end = ndks_boot.freemem[i].end;
-                ndks_boot.freemem[next].start = start + size;
-            }
+            /* There may be leftovers before and after the memory we used. */
+            /* Leave the before leftover in slot i. */
             ndks_boot.freemem[i].end = start;
-            break;
-        } else if (i < ARRAY_SIZE(ndks_boot.freemem)) {
-            ndks_boot.freemem[next] = ndks_boot.freemem[i];
+            /* Shuffle the after leftover up to slot (i + 1), which is unused. */
+            ndks_boot.freemem[i + 1].start = start + size;
```

Comment:
Make this an "atomic" assignment that clearly states we updates the whole slot here and ot parts of it.
```suggestion
            ndks_boot.freemem[i + 1] = (region_t) { 
                .start = start + size;
                .end = ndks_boot.freemem[i].end;
            };
```

Reply:
> for the sake of some code golf.

Sorry, this comment of mine was neither fair nor reasonable.

Reply:
No problem.